### PR TITLE
feat(query): alter table cluster key

### LINKED ISSUE #5719: Feature:  Support alter table cluster key
**Summary**

```
ALTER TABLE <name> CLUSTER BY ( <expr1> [ , <expr2> ... ] )
```

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog

- New Feature

## Related Issues

Fixes #5719

### CODE REVIEW

File: common/meta/app/src/schema/table.rs
Comment (drmingdrmer) [inline common/meta/app/src/schema/table.rs:259]: This method is not a `set`, it is a `push`.
Maybe naming it with `push_cluster_key` would be better.

File: common/proto-conv/tests/it/proto_conv.rs
Comment (drmingdrmer) [inline common/proto-conv/tests/it/proto_conv.rs:253]: These bytes are built with the `v1` proto-conv for compatibility tests.

It should not be modified. The test has to pass without changing these bytes.

**Comment by drmingdrmer** [comment]
Nice shot!

**Comment by drmingdrmer** [comment]
I found there is a flaw in the compatibility:

The updated protobuf `TableMeta` has changed. The protobuf types should be serialized with a new `ver`.
But the data written by a query with new `ver` can not be read by a query with the old `ver`,
when only some of the query-nodes in a cluster are upgraded.

The crate `proto-conv` has to write the `OLDEST_COMPATIBLE_VER` into protobuf message, so that an old query will be able to find out if it is safe to read the data.

https://github.com/datafuselabs/databend/blob/25531a724d38fd242a03155fbfea63e443b68a03/common/proto-conv/src/util.rs#L17-L18

**Comment by drmingdrmer** [comment]
This PR has to merge this patch first:
- #5785

And it'd better increase `VER` to a higher value e.g. `2` to indicate meta changes, after #5785 is merged.

**Comment by zhyass** [comment]
> And it'd better increase `VER` to a higher value e.g. `2` to indicate meta changes, after #5785 is merged.

Do any changes need to be made in this pr, after https://github.com/datafuselabs/databend/pull/5785 is merged?

**Comment by drmingdrmer** [comment]
> > And it'd better increase `VER` to a higher value e.g. `2` to indicate meta changes, after #5785 is merged.
> 
> Do any changes need to be made in this pr, after #5785 is merged?

It would be better to upgrade the `VER` every time `proto-conv` crate changes.
https://github.com/datafuselabs/databend/blob/25531a724d38fd242a03155fbfea63e443b68a03/common/proto-conv/src/util.rs#L17-L18

**But** the data with a new `VER` can not be read by query-node without #5785 

Update the `VER` is **not mandatory** if you do not need to differentiate the version in this PR and before this PR. It is totally OK. :)

**Comment by drmingdrmer** [comment]
> > And it'd better increase `VER` to a higher value e.g. `2` to indicate meta changes, after #5785 is merged.
> 
> Do any changes need to be made in this pr, after #5785 is merged?

Hmm... If you do not upgrade the `VER` you do not need to wait for my PR: #5678 to merge this one. 🤔

File: common/ast/src/ast/statement.rs
Comment (sundy-li) [inline common/ast/src/ast/statement.rs:297]: Add or Alter ?

Comment (zhyass) [reply common/ast/src/ast/statement.rs:297]: > Add or Alter ?

Alter is better

**Comment by zhyass** [comment]
> Hmm... If you do not upgrade the `VER` you do not need to wait for my PR: #5678 to merge this one. 🤔

Got it

**Review by drmingdrmer — APPROVED** [review]

**Review by dantengsky — APPROVED** [review]
LGTM

File: common/meta/app/src/schema/table.rs
Comment (andylokandy) [inline common/meta/app/src/schema/table.rs:179]: `Why are here duplicated `cluster_key` and `cluster_keys`?

Comment (andylokandy) [reply common/meta/app/src/schema/table.rs:179]: cc @zhyass

Comment (zhyass) [reply common/meta/app/src/schema/table.rs:179]: > Why are there duplicated `cluster_key` and `cluster_keys`?

The `cluster_key` records the current cluster key of the table. The `cluster_keys` records all historical cluster keys of the table.

**Pull request merged**
